### ISSUE TITLE
Allow running more than one WML unit test per instance of Wesnoth.

### ISSUE THREAD

**Issue by Pentarctagon** [wesnoth/wesnoth#4535]
Currently on Travis, the WML unit tests can take a significant amount of time to complete since they each individual unit test starts up its own instance of Wesnoth to run that single test.  Given that there are [a decent number of tests](https://github.com/wesnoth/wesnoth/blob/master/wml_test_schedule), it can end up taking 10+ minutes for them to complete, which has caused travis timeouts occasionally.

It is done this way, since currently the WML tests script uses Wesnoth's [return code](https://github.com/wesnoth/wesnoth/blob/1a67d53e7dd22c675e1fcadf963c7474035e1083/run_wml_tests#L196) to determine if the test returned the expected value.

The change would then be to instead allow specifying multiple tests to run, and then outputting the results of each test after the final test runs.

**Comment by stevecotton** [comment]
Of those tests, 160 expect return code zero, leaving 33 that expect non-zero. Supporting running multiple tests only for the ones that are expected to return zero is 80% of the optimization, and probably simpler than handling the non-zero return codes.

**Comment by Pentarctagon** [comment]
I think probably the only required C++ change needed would be to add support for running multiple tests in a row - maybe just a huge list at the command line or read from a config file - and then the existing unit tests could be updated to instead log their output to stdout, use PersistenceWML, etc.

The results could also be directly output into a csv file, I suppose.

**Comment by stevecotton** [comment]
The trouble with the required C++ change is that the logic involves wesnoth.cpp's `do_gameloop` and the way that variables are set and used in calls to game_initialization.cpp. I've had two goes at trying to understand it, but still feel that I'd only be able to hack another code path on top of it instead of understanding the logic of its design. Why does that function call lots of other functions and return if they return false, rather than just calling the one function that should do something?

Looking at @gfgtdf 's 3989199a88e0cc608832faf585a7c3e863fa496d, I get the feeling that he had to hack on top of it rather than refactoring the logic of its design.

**Comment by stevecotton** [comment]
The `run_wml_tests` script sounds like something that should already be a common open-source tool, and I'd prefer to move to an existing tool instead of needing to refactor a Wesnoth-specific script.

However, my search-fu is failing me - does this tool already exist?

**Comment by Pentarctagon** [comment]
Even if not, that script could probably be trimmed down significantly.  The whole thing seems rather over-engineered for what it's currently used for, and unnecessarily difficult to follow.  I'll probably take a stab at it relatively soon, if you don't end up replacing it.

**Comment by CelticMinstrel** [comment]
Also, it'd be cool if it were written in Python instead of shell so that Windows doesn't need its own private logic for running the tests.

**Comment by Pentarctagon** [comment]
Sure, makes sense.

**Comment by stevecotton** [comment]
@Pentarctagon I just pushed a messy WIP Python script

**Comment by gfgtdf** [comment]
this seems actually rather easy to do, from looking at the code it seems like the -u argument already takes a  list of scenarios, all we need to do it is, to instead of of relying on a programm returns value,  parse its output and search for the "<scenario_name>:<fail/pass>" lines.

**Comment by stevecotton** [comment]
To continue after a non-zero result, we need to reset any changes that the failing test made in the engine, such as the `lg::broke_strict()` flag. Resetting a flag is easy, but doing so would add another connection to the already-complex main loop, and I don't know if that will be the only change needed to make all the tests run in a single instance.

By "already-complex", I mean that wesnoth.cpp, game_launcher.cpp, commandline_options.cpp, and title_screen.cpp use several single-instance-with-program-lifecycle objects with the logic of when things get set or reset being hard to follow. I agree entirely with @AI0867's comment that it would be good to use RAII, but I don't volunteer to do the refactoring.

**Comment by gfgtdf** [comment]
i don't think the code in wesnoth.cpp or title_screen.cpp is of any interest for this feature, all they do are handling user input, parse commandline args and  do one-time setups that stay the same durign a campaign run. i'd just say reset that flag at the end of the loop inside game_launcher.cpp::unit_test() and you are good

**Comment by stevecotton** [comment]
We don't know if resetting a flag is the only change needed. For tests that expect something to set the broke-strict flag, it's possible that some residual state will cause subsequent tests to behave as expected (setting the broke-strict flag), even when code being tested is broken (it wouldn't set the broke-strict flag when run as a standalone test).

However, we've now separated several more possible return values for tests. The FAIL (1), FAIL_BY_DEFEAT (7) and PASS_BY_VICTORY (8) results all seem as if they could be turned into bundles.
